src/thor/unidirectional_astar.cc

```diff
@@ -649,50 +652,59 @@ void UnidirectionalAStar<expansion_direction, FORWARD>::SetOrigin(
+   * In this case the edge candidates for both locations are node snapped to B meaning both
+   * locations get both directions of AB again but this time they are on the same node. We can call
+   * this route super-trivial because its not just a single edge its actually a single node. In both
+   * SetOrigin and in SetDestination we try to remove edge candidates that are superfluous as
+   * described above. SetDestination happens first and trivially removes the edge candidate for 2
+   * that travels from B back to A. This means the destination now can only be reached via the edge
+   * traveling from A to B. However that edge candidate of 1 is 100% along that edge and would be
+   * trivially removed, since its a node snap that is arriving at the origin rather than leaving it.
+   * So here in SetOrigin we add a bit more complicated logic to say that if this case occurs, allow
+   * that edge to be used
    */
 
-  // its super trivial if both are node snapped to the same end of the same edge
+  // it's super trivial if both are node snapped to the same end of the same edge
   // note the check for node snapping is in the if below and not in this lambda
-  auto super_trivial = [this](const valhalla::PathEdge& edge) {
-    auto p = destinations_percent_along_.find(edge.graph_id());
-    return p != destinations_percent_along_.end() && edge.percent_along() == p->second;
+  auto super_trivial = [&](const valhalla::PathEdge& edge) {
+    GraphId edgeid(edge.graph_id());
+    auto dests = destinations_.equal_range(edgeid);
+    for (auto it = dests.first; it != dests.second; ++it) {
+      const auto& dest_path_edge = it->second.get();
+      if (edge.percent_along() == dest_path_edge.percent_along()) {
+        return true;
+      }
+    }
+    return false;
   };
 
   // Iterate through edges and add to adjacency list
   for (const auto& edge : origin.correlation().edges()) {
-    // If this is a node snap and we have other candidates that we can skip this unless its the one we
-    // need for super trivial route
+    // If this is a node snap and we have other candidates that we can skip this unless its the one
+    // we need for super trivial route
     if ((FORWARD ? edge.end_node() : edge.begin_node()) && has_other_edges && !super_trivial(edge)) {
       continue;
     }
 
     GraphId edgeid(edge.graph_id());
+    auto percent_along = edge.percent_along();
 
-    // Disallow any user avoid edges if the avoid location is ahead of the origin along the edge
-    if (FORWARD) {
-      if (costing_->AvoidAsOriginEdge(edgeid, edge.percent_along())) {
-        continue;
-      }
+    // Disallow any user avoided edges if the avoid location is behind the destination along the
+    // edge (check has to be done BEFORE we invert the edge if !FORWARD)
+    if (FORWARD ? costing_->AvoidAsOriginEdge(edgeid, percent_along)
+                : costing_->AvoidAsDestinationEdge(edgeid, percent_along)) {
+      continue;
     }
```

Comment:
Hm, not so convinced anymore after all.. For **reverse** this might not do what we'd expect I think. Basically it seems this relies on the edge as it's being traversed by the algorithm, not its opposing. That exact edge might not even be in the `avoid_edges` set if the current mode isn't allowed on it. However, its opposing (the one we actually care about) might've been. That is to say, it's better than before, where we didn't look at avoided edges at all for the reverse tree, but might still need some more fiddling to get it really right, so we should leave a TODO.